### PR TITLE
[inductor] Lower divide by constant as multiplication by reciprocal

### LINKED ISSUE #101039: [torch.compile] the sum of `softmax` isn't `1` on cuda
### 🐛 Describe the bug

`torch.compile` returns the sum of `softmax` isn't equal to `1` on cuda

```py
import torch

torch.manual_seed(420)

class Model(torch.nn.Module):

    def __init__(self):
        super().__init__()
        self.softmax = torch.nn.Softmax(dim=-1)

    def forward(self, query, key):
        y = torch.matmul(query, key.transpose(-2, -1))
        z = y.div(1e-06)
        w = self.softmax(z)
        return w

device = 'cuda'
func = Model().to(device)

query = torch.randn(1, 10, 40).to(device)

key = torch.randn(1, 2, 40).to(device)

jit_func = torch.compile(func)

res1 = func(query, key) # without jit
print(res1)
# tensor([[[0., 1.],
#          [0., 1.],
#          [0., 1.],
#          [0., 1.],
#          [0., 1.],
#          [1., 0.],
#          [1., 0.],
#          [0., 1.],
#          [0., 1.],
#          [0., 1.]]], device='cuda:0')

res2 = jit_func(query, key)
print(res2)
# tensor([[[0.0000, 0.9804],
#          [0.0000, 1.0561],
#          [0.0000, 1.1190],
#          [0.0000, 0.9721],
#          [0.0000, 1.1999],
#          [0.9972, 0.0000],
#          [0.8769, 0.0000],
#          [0.0000, 1.0315],
#          [0.0000, 1.0398],
#          [0.0000, 0.8039]]], device='cuda:0')
```

### Versions

<details>
<summary>Click to expand</summary>

```
Collecting environment information...
PyTorch version: 2.1.0.dev20230503+cu118
Is debug build: False
CUDA used to build PyTorch: 11.8
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.1 LTS (x86_64)
GCC version: (Ubuntu 11.3.0-1ubuntu1~22.04) 11.3.0
Clang version: 14.0.0-1ubuntu1
CMake version: Could not collect
Libc version: glibc-2.35

Python version: 3.9.16 (main, Mar  8 2023, 14:00:05)  [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.19.5-051905-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 11.5.119
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 3060
Nvidia driver version: 510.108.03
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Address sizes:                   46 bits physical, 48 bits virtual
Byte Order:                      Little Endian
CPU(s):                          24
On-line CPU(s) list:             0-23
Vendor ID:                       GenuineIntel
Model name:                      12th Gen Intel(R) Core(TM) i9-12900K
CPU family:                      6
Model:                           151
Thread(s) per core:              2
Core(s) per socket:              16
Socket(s):                       1
Stepping:                        2
CPU max MHz:                     6700.0000
CPU min MHz:                     800.0000
BogoMIPS:                        6374.40
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault cat_l2 invpcid_single cdp_l2 ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid rdt_a rdseed adx smap clflushopt clwb intel_pt sha_ni xsaveopt xsavec xgetbv1 xsaves split_lock_detect avx_vnni dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp hwp_pkg_req hfi umip pku ospke waitpkg gfni vaes vpclmulqdq tme rdpid movdiri movdir64b fsrm md_clear serialize pconfig arch_lbr ibt flush_l1d arch_capabilities
Virtualization:                  VT-x
L1d cache:                       640 KiB (16 instances)
L1i cache:                       768 KiB (16 instances)
L2 cache:                        14 MiB (10 instances)
L3 cache:                        30 MiB (1 instance)
NUMA node(s):                    1
NUMA node0 CPU(s):               0-23
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Mmio stale data:   Not affected
Vulnerability Retbleed:          Not affected
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Enhanced IBRS, IBPB conditional, RSB filling, PBRSB-eIBRS SW sequence
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected

Versions of relevant libraries:
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.24.2
[pip3] pytorch-triton==2.1.0+7d1a95b046
[pip3] torch==2.1.0.dev20230503+cu118
[pip3] torchaudio==2.1.0.dev20230503+cu118
[pip3] torchvision==0.16.0.dev20230503+cu118
[conda] numpy                     1.24.2                   pypi_0    pypi
[conda] pytorch-triton            2.1.0+7d1a95b046          pypi_0    pypi
[conda] torch                     2.1.0.dev20230503+cu118          pypi_0    pypi
[conda] torchaudio                2.1.0.dev20230503+cu118          pypi_0    pypi
[conda] torchvision               0.16.0.dev20230503+cu118          pypi_0    pypi
```

</details>

cc @ezyang @msaroufim @wconstab @ngimel @bdhirsh @anijain2305 @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @soumith

### PR DESCRIPTION
**Pull request by peterbell10**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #122736
* #122915
* #115435
* #122518
* __->__ #121924

Fixes #101039

This lowers division by a constant value to be multipication by reciprocal.
The same optimization is applied in eager mode on CUDA:

https://github.com/pytorch/pytorch/blob/0636c11811e15d1919cdd6cf20cb2d2bed2ee1da/aten/src/ATen/native/cuda/BinaryDivTrueKernel.cu#L36-L38

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

File: torch/_inductor/lowering.py [hunk 1]
```diff
@@ -5339,6 +5342,35 @@ def mul(a, b):
...
+        return get_constant_value(x.unwrap_view())
+    if isinstance(x, ir.Constant):
+        return x
+
+    loader = x.make_loader()
+    if hasattr(x, "inner_fn_args"):
+        inner_fn_args = x.inner_fn_args()
+    else:
+        size = x.get_size()
+        idx = [sympy_index_symbol(f"i{n}") for n, s in enumerate(size)]
```
Comment (lezcano) [inline torch/_inductor/lowering.py:5359]:
```suggestion
        idx = [sympy_index_symbol(f"i{n}") for n in range(len(size))]
```

Comment (lezcano) [reply torch/_inductor/lowering.py:5359]: Even more, I see this pattern in quite a few places. Do we want to expose it in a coherent API all throughout the IR?

File: torch/_inductor/lowering.py [hunk 2]
```diff
@@ -5339,6 +5342,35 @@ def mul(a, b):
         return make_pointwise(fn)(a, b)
 
 
+def get_constant_value(x: ir.IRNode) -> Optional[ir.Constant]:
+    """Try convert an arbitrary IR node into an ir.Constant value"""
+    if isinstance(x, ir.MutableBox):
+        return get_constant_value(x.data)
+    if isinstance(x, ir.BaseView):
+        return get_constant_value(x.unwrap_view())
```
Comment (lezcano) [inline torch/_inductor/lowering.py:5350]: Is this just an optimisation? Could we let the code below handle these two cases?

File: torch/_inductor/ops_handler.py [hunk 3]
```diff
@@ -664,3 +664,18 @@ def count(val):
 
 def _typecheck_OpCounterCSE(h: OpCounterCSE) -> OpsHandler[str]:
     return h
+
+
+class ExtractConstantsHandler(KernelFormatterHandler):
```
Comment (lezcano) [inline torch/_inductor/ops_handler.py:669]: Why do you inherit from `KernelFormatterHandler`?

**Review by lezcano — COMMENTED** [review]
Any reason for not turning this into an actual generic constant folding pass, where anything that's constant is represented by `ir.Constant`?

**Comment by lezcano** [comment]
As a separate point, do we also use this lowering for CPU? A priori, I'd assume we don't

**Comment by peterbell10** [comment]
> Any reason for not turning this into an actual generic constant folding pass, where anything that's constant is represented by `ir.Constant`?

This could certainly be done as an extension, but would be a more complicated change. 

> As a separate point, do we also use this lowering for CPU? A priori, I'd assume we don't

We don't in eager, but this applies the optimization to both CPU and CUDA.

File: torch/_inductor/ops_handler.py (hunk 3 above)
Comment (peterbell10) [reply torch/_inductor/ops_handler.py:669]: This is a workaround for `MockHandler`'s exponential blow-up issue. Basically if you do `x = ops.add(x, x)` repeatedly it expands to a single expression with 2^n elements.

Would be better to add a new ops handler base class that doesn't do any string formatting, but this would be a different change.

File: torch/_inductor/lowering.py (hunk 2 above)
Comment (peterbell10) [reply torch/_inductor/lowering.py:5351]: Yeah, I could just drop these and put everything through `ExtractConstantsHandler` but this avoids creating all the intermediate sympy indexing expressions if they aren't required.

File: torch/_inductor/ops_handler.py (hunk 3 above)
Comment (lezcano) [reply torch/_inductor/ops_handler.py:669]: Do you mind doing that refactoring? This is too hacky otherwise.

**Comment by lezcano** [comment]
>We don't in eager, but this applies the optimization to both CPU and CUDA

I know, but do we want to? If so, we should probably also change CPU eager. cc @jgong5

File: torch/_inductor/lowering.py (hunk 2 above)
Comment (lezcano) [reply torch/_inductor/lowering.py:5351]: Mind leaving a comment?

File: torch/_inductor/lowering.py (hunk 1 above)
Comment (peterbell10) [reply torch/_inductor/lowering.py]: Yeah it would be useful if all IR nodes had `inner_fn_args`.

**Comment by jgong5** [comment]
> I know, but do we want to? If so, we should probably also change CPU eager. cc @jgong5

Yes, we'd better also make the behavior of CPU and CUDA consistent in eager. cc @mingfeima 

@lezcano Perhaps we follow this up after this PR lands?

File: torch/_inductor/ops_handler.py [hunk 4]
```diff
@@ -494,6 +494,38 @@ def libdevice_log(self, x0: T) -> T:
...
+    def masked(mask, body, other) -> None:
+        return None
+
+    @staticmethod
+    def frexp(x) -> Tuple[None, None]:
+        return (None, None)
+
+    @staticmethod
+    def scan(dtypes, combine_fn, values, inits) -> Tuple[None, ...]:
+        return tuple(None for i in range(len(values)))
```
Comment (lezcano) [inline torch/_inductor/ops_handler.py:517]: nit.
```suggestion
        return (None,) * len(values)
```
but sure.

**Review by lezcano — APPROVED** [review]
